torchgeo/samplers/single.py

```diff
@@ -200,17 +201,24 @@ def __init__(
         for hit in self.index.intersection(tuple(self.roi), objects=True):
             bounds = BoundingBox(*hit.bounds)
             if (
-                bounds.maxx - bounds.minx > self.size[1]
-                and bounds.maxy - bounds.miny > self.size[0]
+                bounds.maxx - bounds.minx >= self.size[1]
+                and bounds.maxy - bounds.miny >= self.size[0]
             ):
                 self.hits.append(hit)
 
         self.length: int = 0
         for hit in self.hits:
             bounds = BoundingBox(*hit.bounds)
 
-            rows = int((bounds.maxy - bounds.miny - self.size[0]) // self.stride[0]) + 1
-            cols = int((bounds.maxx - bounds.minx - self.size[1]) // self.stride[1]) + 1
+            # last patch samples outside the bounds
+            rows = math.ceil(
+                (bounds.maxy - bounds.miny - self.size[0] + self.stride[0])
+                / self.stride[0]
+            )
```

Comment:
Going to switch to +1 to match https://theano-pymc.readthedocs.io/en/latest/tutorial/conv_arithmetic.html#no-zero-padding-non-unit-strides and other derivations I've seen. The only difference between this and our implementation is that we replace floor with ceil since we want to sample the last patch.

Reply:
Yep, glad we finally got it merged, sorry it took so long!